node/lib/util/checkout.js

```diff
@@ -39,102 +39,58 @@ const colors  = require("colors");
-     */
-    const checkout = co.wrap(function *(repo, status) {
-        if (status.currentBranchName === branchName) {
-            return;                                                   // RETURN
-        }
-        if (null === (yield GitUtil.findBranch(repo, branchName))) {
-            const head = yield repo.getCommit(status.headCommit);
-            yield repo.createBranch(branchName,
-                                    head,
-                                    0,
-                                    repo.defaultSignature());
-        }
+    yield GitUtil.setHeadHard(metaRepo, commit);
 
-        // Do a force because (a) we've already validated that there are no
-        // changes and (b) it won't change the branch in the meta repo
-        // otherwise.
+    const open = yield SubmoduleUtil.listOpenSubmodules(metaRepo);
+    const names = yield SubmoduleUtil.getSubmoduleNamesForCommit(metaRepo,
+                                                                 commit);
+    const shas = yield SubmoduleUtil.getSubmoduleShasForCommit(metaRepo,
+                                                               names,
+                                                               commit);
 
-        yield repo.checkoutBranch(branchName, {
-            checkoutStrategy: NodeGit.Checkout.STRATEGY.FORCE,
-        });
-    });
+    // checkout submodules
 
-    /**
-     * Validate the specified `repo` and fail using the specified `description`
-     * if `repo` is not in the correct state according to the `create`
-     * parameter.
-     */
-    const validate = co.wrap(function *(repo, description) {
-        const branch = yield GitUtil.findBranch(repo, branchName);
-        if (null !== branch && create === "all") {
-            throw new UserError(`
-${description} already has a branch named ${colors.red(branchName)}.`);
-        }
-        if (null === branch && create === "none") {
-            throw new UserError(`
-${description} does not have a branch named ${colors.red(branchName)}.`);
-        }
-    });
+    yield open.map(co.wrap(function *(name) {
+        // Open repo but not alive on this commit.
 
-    // Skip validation if `"some" === create` because every configuration is
-    // valid in that case.
+        if (!(name in shas)) {
+            return; // RETURN
+        }
 
-    if ("some" !== create) {
-        let validators = openSubNames.map(
-            (sub, i) => validate(openSubRepos[i], sub)
-        );
-        validators.push(validate(metaRepo, "The meta-repo"));
-        yield validators;
-    }
+        const repo = yield SubmoduleUtil.getRepo(metaRepo, name);
+        const sha = shas[name];
+        yield GitUtil.fetchSha(repo, sha);
+        const commit = yield repo.getCommit(sha);
+        yield GitUtil.setHeadHard(repo, commit);
+    }));
 
-    // TODO: I believe there is a bug in nodegit/libgit2 somewhere that results
-    // in a crash 
+    // if 'committish' is a branch,  make it current
```

Comment:
It will put you in a detached head state if you don't specify a branch. By this comment, I mean that when commitish is also the name of a branch, we'll make it the current branch.